File: windows/src/main/csharp/ch/cyberduck/ui/controller/DefaultBookmarkController.cs
// 
// Copyright (c) 2010-2017 Yves Langisch. All rights reserved.
// http://cyberduck.io/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// feedback@cyberduck.io
// 

using System.Collections.Generic;
using ch.cyberduck.core;
using ch.cyberduck.core.exception;
using Ch.Cyberduck.Core;
using org.apache.logging.log4j;
using org.apache.commons.lang3;

namespace Ch.Cyberduck.Ui.Controller
{
    public sealed class DefaultBookmarkController : BookmarkController<IBookmarkView>
    {
        private static readonly Logger Log = LogManager.getLogger(typeof(DefaultBookmarkController).FullName);

        public DefaultBookmarkController(Host bookmark) : base(bookmark)
        {
            View.ChangedPasswordEvent += View_ChangedPasswordEvent;
        }

        private void View_ChangedPasswordEvent()
        {
            if (_options.keychain() && _options.password())
            {
                if (string.IsNullOrWhiteSpace(_host.getHostname()))
                {
                    return;
                }
                if (string.IsNullOrWhiteSpace(_host.getCredentials().getUsername()))
                {
                    return;
                }
                if (string.IsNullOrWhiteSpace(View.Password))
                {
                    return;
                }
                try
                {
                    _host.getCredentials().setPassword(View.Password);
                    PasswordStoreFactory.get().save(_host);
                }
                catch (LocalAccessDeniedException e)
                {
                    Log.error($"Failure saving credentials for ${_host} in keychain. ${e.getDetail()}");
                }
            }
        }

        public static class Factory
        {
            private static readonly IDictionary<Host, DefaultBookmarkController> Open =
                new Dictionary<Host, DefaultBookmarkController>();

            public static DefaultBookmarkController Create(Host host)
            {
                DefaultBookmarkController c;
                if (Open.TryGetValue(host, out c))
                {
                    c.Update();
                    return c;
                }

                c = new DefaultBookmarkController(host);
                c.View.ViewClosedEvent += () => Open.Remove(host);
                Open.Add(host, c);
                return c;
            }
        }
    }
}

File: osx/src/main/java/ch/cyberduck/ui/cocoa/controller/DefaultBookmarkController.java
package ch.cyberduck.ui.cocoa.controller;

/*
 * Copyright (c) 2002-2016 iterate GmbH. All rights reserved.
 * https://cyberduck.io/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

import ch.cyberduck.binding.application.NSButton;
import ch.cyberduck.binding.application.NSPopUpButton;
import ch.cyberduck.binding.application.NSSecureTextField;
import ch.cyberduck.binding.application.NSTextField;
import ch.cyberduck.core.Host;
import ch.cyberduck.core.LoginOptions;
import ch.cyberduck.core.Protocol;
import ch.cyberduck.core.preferences.HostPreferencesFactory;
import ch.cyberduck.ui.LoginInputValidator;

public abstract class DefaultBookmarkController extends BookmarkController {

    private final Host bookmark;
    private final LoginOptions options;

    public DefaultBookmarkController(final Host bookmark) {
        this(bookmark, new LoginOptions(bookmark.getProtocol()));
    }

    public DefaultBookmarkController(final Host bookmark, final LoginOptions options) {
        super(bookmark, new LoginInputValidator(bookmark, options), options);
        this.bookmark = bookmark;
        this.options = options;
    }

    public DefaultBookmarkController(final Host bookmark, final LoginInputValidator validator, final LoginOptions options) {
        super(bookmark, validator, options);
        this.bookmark = bookmark;
        this.options = options;
    }

    @Override
    public void setProtocolPopup(final NSPopUpButton button) {
        button.superview().setHidden(!HostPreferencesFactory.get(bookmark).getBoolean("bookmark.protocol.configurable"));
        super.setProtocolPopup(button);
    }

    @Override
    public void setPasswordField(final NSSecureTextField field) {
        this.addObserver(host -> field.superview().setHidden(!options.password));
        super.setPasswordField(field);
    }

    @Override
    public void setCertificatePopup(final NSPopUpButton button) {
        this.addObserver(host -> button.superview().setHidden(!options.certificate));
        super.setCertificatePopup(button);
    }

    @Override
    public void setTimezonePopup(final NSPopUpButton button) {
        this.addObserver(host -> button.superview().setHidden(host.getProtocol().isUTCTimezone()));
        super.setTimezonePopup(button);
    }

    @Override
    public void setEncodingPopup(final NSPopUpButton button) {
        this.addObserver(host -> button.superview().setHidden(!host.getProtocol().isEncodingConfigurable()));
        super.setEncodingPopup(button);
    }

    @Override
    public void setFtpModePopup(final NSPopUpButton button) {
        this.addObserver(host -> button.superview().setHidden(!(host.getProtocol().getType() == Protocol.Type.ftp)));
        super.setFtpModePopup(button);
    }

    @Override
    public void setHostField(final NSTextField field) {
        this.addObserver(host -> field.superview().superview().setHidden(!host.getProtocol().isHostnameConfigurable() && !host.getProtocol().isPortConfigurable()));
        super.setHostField(field);
    }

    @Override
    public void setPortField(final NSTextField field) {
        this.addObserver(host -> field.superview().superview().setHidden(!host.getProtocol().isHostnameConfigurable() && !host.getProtocol().isPortConfigurable()));
        super.setPortField(field);
    }

    @Override
    public void setPathField(final NSTextField field) {
        this.addObserver(host -> field.superview().setHidden(!host.getProtocol().isPathConfigurable()));
        super.setPathField(field);
    }

    @Override
    public void setUsernameField(final NSTextField field) {
        this.addObserver(host -> field.superview().setHidden(!host.getProtocol().isUsernameConfigurable()));
        super.setUsernameField(field);
    }

    @Override
    public void setAnonymousCheckbox(final NSButton button) {
        this.addObserver(host -> button.superview().setHidden(!host.getProtocol().isAnonymousConfigurable()));
        super.setAnonymousCheckbox(button);
    }

    @Override
    public void setPrivateKeyPopup(final NSPopUpButton button) {
        this.addObserver(host -> button.superview().setHidden(!host.getProtocol().isPrivateKeyConfigurable()));
        super.setPrivateKeyPopup(button);
    }
}
